test/src/specs/tx_pool/dao.rs

```diff
@@ -0,0 +1,415 @@
+use ckb_resource::CODE_HASH_DAO;
+use ckb_test_chain_utils::always_success_cell;
+use numext_fixed_hash::H256;
+
+const SYSTEM_CELL_ALWAYS_SUCCESS_INDEX: u32 = 1;
+const SYSTEM_CELL_DAO_INDEX: u32 = 3;
+const WITHDRAW_WINDOW_LEFT: u64 = 10;
+const WITHDRAW_HEADER_INDEX: u8 = 0;
+
+pub struct DepositDAO;
+
+impl Spec for DepositDAO {
+    crate::name!("deposit_dao");
+
+    fn run(&self, net: Net) {
+        let node0 = &net.nodes[0];
+        node0.generate_blocks(2);
+
+        // Deposit DAO
+        {
+            let transaction = deposit_dao_transaction(node0);
+            ensure_committed(node0, &transaction);
+        }
+
+        // Deposit DAO without specifying `block_hash` within input
+        {
+            // deposit dao transaction without `block_hash`
+            let transaction = {
+                let out_point_without_block_hash = {
+                    let tx_hash = node0.get_tip_block().transactions()[0].hash().to_owned();
+                    OutPoint::new_cell(tx_hash, 0)
+                };
+                TransactionBuilder::from_transaction(deposit_dao_transaction(node0))
+                    .inputs_clear()
+                    .input(CellInput::new(out_point_without_block_hash, 0))
+                    .build()
+            };
+            ensure_committed(node0, &transaction);
+        }
+    }
+}
+
+pub struct WithdrawDAO;
+
+impl Spec for WithdrawDAO {
+    crate::name!("withdraw_dao");
+
+    fn run(&self, net: Net) {
+        let node0 = &net.nodes[0];
+        node0.generate_blocks(2);
+
+        let deposited = {
+            let transaction = deposit_dao_transaction(node0);
+            ensure_committed(node0, &transaction)
+        };
+        let transaction = withdraw_dao_transaction(node0, deposited.clone());
+        ensure_committed(node0, &transaction);
+    }
+}
+
+pub struct WithdrawAndDepositDAOWithinSameTx;
+
+impl Spec for WithdrawAndDepositDAOWithinSameTx {
+    crate::name!("withdraw_and_deposit_dao_within_same_tx");
+
+    fn run(&self, net: Net) {
+        let node0 = &net.nodes[0];
+        node0.generate_blocks(2);
+
+        let mut deposited = {
+            let transaction = deposit_dao_transaction(node0);
+            ensure_committed(node0, &transaction)
+        };
+        for _ in 0..5 {
+            let transaction = {
+                let transaction = withdraw_dao_transaction(node0, deposited.clone());
+                let outputs: Vec<_> = transaction
+                    .outputs()
+                    .iter()
+                    .cloned()
+                    .map(|mut cell_output| {
+                        cell_output.type_ = Some(deposit_dao_script());
+                        cell_output
+                    })
+                    .collect();
+                TransactionBuilder::from_transaction(transaction)
+                    .outputs_clear()
+                    .outputs(outputs)
+                    .build()
+            };
+            // TODO compare the reward
```

Comment:
Add `node0.generate_blocks(20);` otherwise the tx is not  maturity.

Reply:
`ensure_committed` will do it.